File: chapters/sorting_searching/bogo/code/c/bogo_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

bool is_sorted(int *array, int n) {
    while (--n >= 1) {
        if (array[n] < array[n - 1]) {
            return false;
        }
    }

    return true;
}

void shuffle(int *array, int n) {
    for (int i = 0; i < n; ++i) {
        int t = array[i];
        int r = rand() % n;
        array[i] = array[r];
        array[r] = t;
    }
}

void bogo_sort(int *array, int n) {
    while (!is_sorted(array, n)) {
        shuffle(array, n);
    }
}

int main() {
    int array[] = { 1, 3654, 78, 654, -234, -12, 4, 3, -6, -100};

    printf("Unsorted array:\n");
    for (int i = 0; i < 10; ++i) {
        printf("%d ", array[i]);
    }
    printf("\n\n");

    bogo_sort(array, 10);

    printf("Sorted array:\n");
    for (int i = 0; i < 10; ++i) {
        printf("%d ", array[i]);
    }
    printf("\n");
}

File: chapters/sorting_searching/bogo/code/rust/bogosort.rs
// Submitted by jess 3jane

extern crate rand;

fn shuffle(arr : &mut Vec<i32>) {
    for i in 0..arr.len() {
        let t = arr[i];
        let r : usize = rand::random::<usize>() % arr.len();
        arr[i] = arr[r];
        arr[r] = t;
    }
}

fn is_sorted(arr : &Vec<i32>) -> bool {
    for i in 1..arr.len() {
        if arr[i-1] > arr[i] {
            return false;
        }
    }
    true
}

fn bogo_sort(arr : &mut Vec<i32>) {
    while !is_sorted(arr) {
        shuffle(arr)
    }
}

fn main() {
    let mut v = vec!(1,2,3,4,5);
    shuffle(&mut v);
    println!("Original array: {:?}", v);
    bogo_sort(&mut v);
    println!("Sorted array: {:?}", v);
}
